merlin/models/tf/outputs/contrastive.py

```diff
@@ -248,7 +265,13 @@ def call_contrastive(self, inputs, features, targets, training=False, testing=Fa
         ):
             negative = negative.with_embedding(self.embedding_lookup(negative.id))
 
-        return self.outputs(query_embedding, positive, negative)
+        logits = self.outputs(query_embedding, positive, negative)
+        if is_ragged:
+            logits.copy_with_updates(
+                outputs=original_query_embedding.with_flat_values(logits.outputs),
+                targets=original_target.with_flat_values(logits.targets),
```

Comment:
This with_flat_values is very useful! Seems faster than rebuilding the full ragged tensor.